### DISCUSSION TITLE
GraphHopper (offline) curvy routing

### DISCUSSION THREAD

**Discussion by devemux86** [devemux86/cruiser#166] [Q&A] [answered]
_"Prefers routes with lots of curves for enjoyable rides"_

Combined with the ignore barriers option (see https://github.com/devemux86/cruiser/discussions/56#discussioncomment-5897212),
it can produce interesting results for a basic **offline** use.

It can be enabled for the motorcycle profile in the next update.

**Comment by Ciutnik** [comment]
Does this mean that by sending an appropriate query to Graphhopper-offline you can have an impact on the designated route? I mean not just simple avoidance of highways or unpaved roads, but the level of avoidance/preference of particular road categories? 😳
If so, we have an open road to the well-known and loved K1/K2 algorithm?

**Reply by devemux86** [reply]
Currently it is not so advanced and without options.
But it's certainly not the usual fastest route that most motor vehicle routers calculate.

I'm exploring our options with BRouter and GraphHopper and we'll see how it goes.
Important thing is also to be able to work **offline**.

**Comment by Ciutnik** [comment]
> Important thing is also to be able to work offline.

Absolutely yes! And by the way, in my opinion GraphHopper has the advantage over BRouter because it returns street names.
Thank you very much for your commitment to curvy routing 👍

**Reply by devemux86** [reply]
> it returns street names

And speed limits.

However, BRouter is easier to customize without writing code.
Each router has advantages.

We will see, many new things are coming to Cruiser in the near future. :slightly_smiling_face:

**Comment by devemux86** [comment]
Coming in next Cruiser:

<img src="https://github.com/devemux86/cruiser/assets/3484020/f855c630-f7b4-479b-aa48-c652d8929c34" width="50%"/>

**Comment by Leeb64** [comment]
A general question about Graphhopper:
I live in Germany near the French border.
Planning with Graphhopper in D works!
Planning in F doesn't work.
I know that Graphhopper cannot plan across the border...but why can't I plan anything in F?
Of course I downloaded the graphs!
Surely it can't be that you have to be in the country where you want to plan a route?

**Reply by devemux86** [reply]
GraphHopper (offline) requires download and activation of the country we want to do route planning.

Have you activated France graph before do route planning inside France?
(tap the country and activate it, the active has blue color)

<img src="https://github.com/devemux86/cruiser/assets/3484020/390e1753-8aa3-444b-b0c8-54c759685ceb" width="50%"/>

**Reply by Leeb64** [reply]
Learned something again!

**Reply by devemux86** [reply]
It should display the message: "GraphHopper: Waypoint N outside supported area"
when we place waypoints outside the active graph.

A similar message is also displayed by BRouter.

**Comment by devemux86** [comment, marked as answer]
GraphHopper (offline) curvy routing (first implementation) is available in [Cruiser 3.0.20](https://play.google.com/store/apps/details?id=gr.talent.cruiser).

Available in "Settings | Routing | GraphHopper | Curvy route".

_Update your GraphHopper downloads!_

**Comment by Ciutnik** [comment]
Thank you very much for the new features in Cruiser! 👍
Unfortunately, setting a GraphHopper offline route of about 70 km (not too long) ends with a notification „No path found due to maximum nodes exceeded 1000000”. This is probably a limitation of the router that cannot be easily circumvented.

**Reply by devemux86** [reply]
@Ciutnik probably depends on the area and density of road network.
e.g. in my recent tests in Scotland, it worked with even 400+ km routes.

Can you try to _add extra via / shaping points_ in the middle of the route?
This is the recommended way in all routers: BRouter, GraphHopper, Kurviger, ...

**Reply by devemux86** [reply]
It is with the "Curvy route" setting, right?
I can see it with some routes (where it runs without it).
And it works if we add a waypoint somewhere in the route.

Curvy route calculation on mobile is certainly possible, but more difficult.
That's why we better use more waypoints with such routes, to split the calculations.

**Reply by Ciutnik** [reply]
> It is with the "Curvy route" setting, right?

I already know! It's not the "Curvy route" that causes error but "Alternative routes = 2". Leaving "= 1" the route is mapped out quickly and easily, albeit not very curvy ;-)

**Reply by devemux86** [reply]
> It's not the "Curvy route" that causes error but "Alternative routes = 2". Leaving "= 1" the route is mapped out quickly and easily

Alternative routes > 1 also makes the calculation more complicated, even for normal routing! :slightly_smiling_face:

**Reply by devemux86** [reply]
> albeit not very curvy

It is a _first implementation_ working **offline** and it depends on the road network of the area.
Making tests with the curvy option on or off, it _tries to avoid_ motorways, etc.

Let's see if there is interest to invest time for more exploration and improvements. :slightly_smiling_face:

**Reply by MotoUKRider** [reply]
I've just seen this "nodes" error message when plotting a simple A-to-B trip of around 170-miles.  Not using "curvy" or 2 alternative routes - just the single one.  I can replicate this on several long journeys up a main motorway here in the UK.

BRouter does it without issue.

**Reply by devemux86** [reply]
GraphHopper has a related parameter:
_"To prevent long running routing queries..."_
_"You can define the maximum visited nodes when routing. This may result in not found connections if there is no connection between two points within the given visited nodes. The default is Integer.MAX_VALUE."_

We can use the "unlimited" default (or increase the recommended 1000000) in the next Cruiser update and we will see how it works.
Also we can change the message to better explain that 2 waypoints are too far away and need to add another waypoint in the route.

**Reply by MotoUKRider** [reply]
Interesting - this was a 3-hour drive of 173-miles, mostly "big" roads but obviously lots of turns at roundabouts and such.
So, does this mean that GraphHopper isn't a good routing engine for long-distance A-to-B travel?

**Reply by devemux86** [reply]
@MotoUKRider 

It depends on the options and the area, also **curvy** or **alternative** routes require more calculations.
That's why it is recommended to _avoid the alternative routes_.

I removed the limit, you can test it with desktop Cruiser 3.0.21 (Beta): https://github.com/devemux86/cruiser/discussions/177#discussioncomment-5986159
It will also be available in the next Android Cruiser update.

**Comment by 0709wiwiwi** [comment]
Windows Cruiser desktop. 
Cruiser version 3.0.20 
GraphHopper graph Belgium recent download.
![Invalid routing data](https://github.com/devemux86/cruiser/assets/54167201/1e4f032a-c284-4d80-921a-0f155d50e727)
By Cruiser  -version 3.0.18 ok.

Valhalla by route type short.
![Valhalla](https://github.com/devemux86/cruiser/assets/54167201/07e34657-aeae-4079-884a-a66697e0259f)

**Reply by devemux86** [reply]
@0709wiwiwi please update the GraphHopper downloads.
(also mentioned [above](https://github.com/devemux86/cruiser/discussions/166#discussioncomment-5965184))

All graphs were updated in the past few days.
I tested Belgium and it works properly. :slightly_smiling_face:

**Reply by 0709wiwiwi** [reply]
Is a new download last hour.
http://ftp-stud.hs-esslingen.de/pub/Mirrors/download.mapsforge.org/graphs/

**Reply by devemux86** [reply]
> Is a new download last hour.

Then it is probably a not valid download or problem in unzip?
I downloaded exactly the same graph and it is loaded and works.

**Reply by devemux86** [reply]
> Valhalla by route type short.

@0709wiwiwi thanks for the report! :+1: 

Valhalla short type will work again properly in the next update. :slightly_smiling_face:

**Reply by devemux86** [reply]
Did you extract the zip on top of the old folder?

(Android always extracts the graphs in new clean folders)

**Reply by devemux86** [reply]
Valhalla short type should be fixed in [Cruiser 3.0.21](https://play.google.com/store/apps/details?id=gr.talent.cruiser).

**Comment by 0709wiwiwi** [comment]
Still again that fault.  Pc nr 2 same problem.
So better someone else will test and report. 
Need a pause now.

**Comment by Ciutnik** [comment]
@0709wiwiwi In Cruiser, I downloaded the GraphHopper data, found your place in Belgium and set the route. Looks OK.

![Screenshot_20230522_205221_Cruiser](https://github.com/devemux86/cruiser/assets/122647782/8d89e0bf-0657-4daa-8de8-2e76fa393b99)

**Reply by 0709wiwiwi** [reply]
Thanks for the test. I continue to struggle with the last install. But still keep losing that battle though.
[GH offline graph failure video](https://youtu.be/bDlhkiAqxr8url)

**Reply by devemux86** [reply]
@0709wiwiwi 

On the desktop, selecting a folder (with Java) is done by **enter** the folder and press OK.
(in the video you seem to stay on the parent folder - without routing data)

**Reply by 0709wiwiwi** [reply]
I activate this exactly the same way as with the old versions where it always works perfectly.
![MWSnap002](https://github.com/devemux86/cruiser/assets/54167201/e3db6021-c2a8-4482-b4d9-5ea897945a22)
![MWSnap003](https://github.com/devemux86/cruiser/assets/54167201/578f6402-c4f9-4a67-98e5-cef78e67a79c)


Could there be "incompatible" residues left over from the old versions both in Cruiser or the graphs ?

**Reply by devemux86** [reply]
In the video you were not inside the belgium folder when you clicked OK.
Cruiser works as it always has. :slightly_smiling_face: 

---

_@0709wiwiwi please create a new topic for the desktop.
This discussion is for the GraphHopper (offline) curvy routing._

**Reply by devemux86** [reply]
@0709wiwiwi please see:

- https://github.com/devemux86/cruiser/discussions/177

**Comment by Leeb64** [comment]
I compared the Graphhopper curvy profile to the BRouter mopped profile=>almost similar results.
Disadvantage: Villages are not bypassed although there is a bypass. Annoying.
I muddled around a motorcycle profile for BRouter.
It reacts strongly to the set speed.
Villages are bypassed from 75 km/h. Still preferred small roads.
[car-eco-cruiser.brf.txt](https://github.com/devemux86/cruiser/files/11543703/car-eco-cruiser.brf.txt)

**Reply by DolomitesTom** [reply]
Villages are bypassed from 75 kilometers upwards or downwards? And how is this influenced by the maximum speed of a road?

**Comment by WolfgangZ42** [comment]
I wanted to test the GraphHopper "curvy" version - within Germany after insanely long calculation time aborted, even with only one alternative route. Then I switched to Austria (Achensee to Hochosterwitz). Here the result:

<img width="600" src="https://github.com/devemux86/cruiser/assets/121806238/58597ebd-bb32-4d81-ab5d-fd1189f50247"/> 

Purple=BRouter primary, Gray=BRouter alternatives, Red=GraphHopper curvy, Blue=Kurviger curvy.

Lot of alternatives ... 😉

**Reply by devemux86** [reply]
> Lot of alternatives ... :wink: 

Indeed, more options are always nice!
Even better if they can work offline. :slightly_smiling_face: 

In GraphHopper (offline), it is better to not enable the alternatives for long routes.
Better to use more waypoints, so the route calculation can be split into segments.
